File: java-migration/backend/portfolio-service/src/main/java/com/portfolio/model/entity/Transaction.java
package com.portfolio.model.entity;

import com.portfolio.model.enums.TransactionStatus;
import com.portfolio.model.enums.TransactionType;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import jakarta.validation.constraints.Size;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.time.LocalTime;

@Entity
@Table(name = "transactions")
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Transaction {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "transaction_date")
    private LocalDate transactionDate;

    @Column(name = "transaction_time")
    private LocalTime transactionTime;

    @NotBlank
    @Size(max = 8)
    @Column(name = "portfolio_id", nullable = false, length = 8)
    private String portfolioId;

    @Size(max = 6)
    @Column(name = "sequence_no", length = 6)
    private String sequenceNo;

    @Size(max = 10)
    @Column(name = "investment_id", length = 10)
    private String investmentId;

    @NotNull
    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false)
    private TransactionType type;

    @NotNull
    @Positive
    @Column(name = "quantity", precision = 15, scale = 4, nullable = false)
    private BigDecimal quantity;

    @Column(name = "price", precision = 15, scale = 4)
    private BigDecimal price;

    @Column(name = "amount", precision = 15, scale = 2)
    private BigDecimal amount;

    @Size(max = 3)
    @Column(name = "currency", length = 3)
    private String currency;

    @Enumerated(EnumType.STRING)
    @Column(name = "status")
    private TransactionStatus status;

    @Column(name = "process_date")
    private LocalDateTime processDate;

    @Size(max = 8)
    @Column(name = "process_user", length = 8)
    private String processUser;

    @PrePersist
    protected void onCreate() {
        if (transactionDate == null) {
            transactionDate = LocalDate.now();
        }
        if (transactionTime == null) {
            transactionTime = LocalTime.now();
        }
        if (status == null) {
            status = TransactionStatus.PENDING;
        }
        if (currency == null) {
            currency = "USD";
        }
    }
}

File: backend/schemas/transaction.py
from pydantic import BaseModel, Field
from datetime import date, time, datetime
from typing import Optional


class TransactionCreate(BaseModel):
    portfolio_id: str = Field(..., min_length=8, max_length=8)
    investment_id: str = Field(..., min_length=1, max_length=10)
    transaction_type: str = Field(..., pattern=r"^(BU|SL|TR|FE)$")
    quantity: float = Field(..., gt=0)
    price: float = Field(..., gt=0)
    currency: str = Field(default="USD", min_length=3, max_length=3)


class TransactionStatusUpdate(BaseModel):
    status: str = Field(..., pattern=r"^[PDFR]$")


class TransactionResponse(BaseModel):
    transaction_id: str
    portfolio_id: str
    investment_id: str
    transaction_date: date
    transaction_time: time
    sequence_no: str
    transaction_type: str
    quantity: float
    price: float
    amount: float
    currency: str
    status: str
    process_date: Optional[datetime] = None
    process_user: str
    created_at: datetime

    class Config:
        from_attributes = True


class TransactionListResponse(BaseModel):
    transactions: list[TransactionResponse]
    total: int
